stripe/src/main/java/com/stripe/android/CustomerSession.java

```diff
@@ -0,0 +1,299 @@
+        if (mStripeApiProxy != null) {
+            runnable.run();
+            return;
+        }
+
+        mThreadPoolExecutor.execute(runnable);
+    }
+
+    void onKeyUpdated(String rawKey) {
+        EphemeralKey key = EphemeralKey.fromString(rawKey);
+        mEphemeralKey = key == null ? EphemeralKey.getEmptyKey() : key;
+        updateCustomerIfNecessary(mEphemeralKey, mProxyNowCalendar);
+    }
+
+    void onKeyError(int code, @Nullable String errorMessage) {
+        // Not yet handling this case
+    }
+
+    private void initializeMainThreadHandler() {
+        mUiThreadHandler = new Handler(Looper.getMainLooper()) {
+            @Override
+            public void handleMessage(Message msg) {
+                super.handleMessage(msg);
+
+                Object messageObject = msg.obj;
+
+                switch (msg.what) {
+                    case CUSTOMER_RETRIEVED:
+                        if (messageObject instanceof Customer) {
+                            mCustomer = (Customer) messageObject;
+                        }
+                        break;
+                }
+            }
+        };
+    }
+
+    private void initializeThreadPoolExecutor() {
+        mThreadPoolExecutor = new ThreadPoolExecutor(
+                THREAD_POOL_SIZE,
+                THREAD_POOL_SIZE,
+                KEEP_ALIVE_TIME,
+                KEEP_ALIVE_TIME_UNIT,
+                mNetworkQueue);
+    }
+
+    /**
+     * Calls the Stripe API (or a test proxy) to fetch a customer. If the provided key is expired,
+     * this method <b>does not</b> update the key.
+     * Use {@link #updateCustomerIfNecessary(EphemeralKey, Calendar)} to validate the key
+     * before refreshing the customer.
+     *
+     * @param key the {@link EphemeralKey} used for this access
+     * @param proxy a {@link StripeApiProxy} to intercept calls to the real servers
+     * @return a {@link Customer} if one can be found with this key, or {@code null} if one cannot.
+     */
+    @Nullable
+    static Customer retrieveCustomerWithKey(
+            @NonNull EphemeralKey key,
+            @Nullable StripeApiProxy proxy,
+            @Nullable Calendar nowCalendar) {
+        Customer customer = null;
+        if (isTimeInPast(key.getExpires(), nowCalendar)) {
+            return null;
+        }
+
+        try {
+            if (proxy != null) {
+                return proxy.retrieveCustomerWithKey(key.getCustomerId(), key.getSecret());
+            }
+            customer = StripeApiHandler.retrieveCustomerWithKey(
+                    key.getCustomerId(),
+                    key.getSecret());
+        } catch (InvalidRequestException invalidException) {
+            // Then the key is invalid
+        } catch (StripeException stripeException) {
+            Log.e(CustomerSession.class.getName(), stripeException.getMessage());
+        }
+        return customer;
+    }
+
+    static boolean isTimeInPast(long milliSeconds, @Nullable Calendar nowCalendar) {
```

Comment:
This method has no idea whether or not it's addressing a key, though -- isTimeWithinDeltaOfPresent(long time, long delta, Calendar debugCalendar) might make it more useful, since i'll need to check on the age of the customer in the future